lightning/src/ln/channel.rs

```diff
@@ -3460,76 +3645,105 @@ impl<Signer: Sign> Channel<Signer> {
-				assert!(self.shutdown_scriptpubkey.is_some());
-				let tx_weight = self.get_closing_transaction_weight(Some(&self.get_closing_scriptpubkey()), Some(self.counterparty_shutdown_scriptpubkey.as_ref().unwrap()));
-				let (closing_tx, used_total_fee) = self.build_closing_transaction($new_feerate as u64 * tx_weight / 1000, false);
+		self.calculate_closing_feerate_limits(fee_estimator);
+		let (our_min_fee, our_max_fee) = self.closing_fee_limits.unwrap();
+
+		macro_rules! propose_new_fee {
+			($new_fee: expr) => {
+				let (mut tx, used_fee) = if $new_fee == msg.fee_satoshis {
+					(closing_tx, $new_fee)
+				} else {
+					self.build_closing_transaction($new_fee, false)
+				};
+
 				let sig = self.holder_signer
-					.sign_closing_transaction(&closing_tx, &self.secp_ctx)
+					.sign_closing_transaction(&tx, &self.secp_ctx)
 					.map_err(|_| ChannelError::Close("External signer refused to sign closing transaction".to_owned()))?;
-				assert!(closing_tx.get_weight() as u64 <= tx_weight);
-				self.last_sent_closing_fee = Some(($new_feerate, used_total_fee, sig.clone()));
+
+				let signed_tx = if $new_fee == msg.fee_satoshis {
+					self.channel_state = ChannelState::ShutdownComplete as u32;
+					self.update_time_counter += 1;
+					self.build_signed_closing_transaction(&mut tx, &msg.signature, &sig);
+					Some(tx)
+				} else { None };
+
+				self.last_sent_closing_fee = Some((used_fee, sig.clone()));
 				return Ok((Some(msgs::ClosingSigned {
 					channel_id: self.channel_id,
-					fee_satoshis: used_total_fee,
+					fee_satoshis: used_fee,
 					signature: sig,
-				}), None))
+					fee_range: Some(msgs::ClosingSignedFeeRange {
+						min_fee_satoshis: our_min_fee,
+						max_fee_satoshis: our_max_fee,
+					}),
+				}), signed_tx))
 			}
 		}
 
-		let mut min_feerate = 253;
-		if self.is_outbound() {
-			let max_feerate = fee_estimator.get_est_sat_per_1000_weight(ConfirmationTarget::Normal);
-			if (msg.fee_satoshis as u64) > max_feerate as u64 * closing_tx_max_weight / 1000 {
-				if let Some((last_feerate, _, _)) = self.last_sent_closing_fee {
-					if max_feerate <= last_feerate {
-						return Err(ChannelError::Close(format!("Unable to come to consensus about closing feerate, remote wanted something higher ({}) than our Normal feerate ({})", last_feerate, max_feerate)));
-					}
+		if let Some(msgs::ClosingSignedFeeRange { min_fee_satoshis, max_fee_satoshis }) = msg.fee_range {
+			if msg.fee_satoshis < min_fee_satoshis || msg.fee_satoshis > max_fee_satoshis {
+				return Err(ChannelError::Close(format!("Peer sent a bogus closing_signed - suggested fee of {} sat was not in their desired range of {} sat - {} sat", msg.fee_satoshis, min_fee_satoshis, max_fee_satoshis)));
+			}
+			if max_fee_satoshis < our_min_fee {
+				return Err(ChannelError::Warn(format!("Unable to come to consensus about closing feerate, remote's max fee ({} sat) was smaller than our min fee ({} sat)", max_fee_satoshis, our_min_fee)));
```

Comment:
To check my understanding -- if we reach this case, is the onus on the counterparty to send a new fee_range?

Reply:
Yep, we (will) tell the counterparty we can't agree to their current fee range, and either they respond with a better one or we force-close after a time out.